### DISCUSSION TITLE
issue: Select "multiple" options in latest version not working using controller

### DISCUSSION THREAD

**Discussion by aruljayaraj** [react-hook-form/react-hook-form#7706] [Q&A] [answered]
### Version Number

7.25.3

### Codesandbox/Expo snack

https://codesandbox.io/s/optimistic-hoover-zjotj?file=/src/App.tsx

### Steps to reproduce

1. Go to the given Sandbox link
2. Select the colors from the dropdown
3. check the console on the onChange and onSubmit event.
4. Selected values not displayed onSubmit.

### Expected behaviour

I've recently upgraded the **react-hook-form** from **6.14.1 to 7.25.3**(latest) in my project and also updated the **react** version from **16.13.0 to 17.0.2**
I would like to get the selected colors during onSubmit. I don't know how to get that values in the recent version. It was working in the lower version of react-hook-form.

### What browsers are you seeing the problem on?

Firefox, Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment, marked as answer]
https://react-hook-form.com/api/usecontroller/controller

- missing defaultValues 
- selected.target.selectedOptions instead e.target.value
- missing value assigned to the Select component

**Reply by Moshyfawn** [reply]
For example:
```js
<select
  multiple
  onChange={(event) =>
    onChange(
      Array.from(event.target.selectedOptions).map(
        (selectedOption) => selectedOption.value
      )
    )
  }
  value={value}
/>
```

**Reply by aruljayaraj** [reply]
@Moshyfawn It's working fine. Thanks a lot.

**Reply by Moshyfawn** [reply]
@aruljayaraj, pls, take note, that in order to properly set a default value on a controlled input element, you should add
```js
useForm({
  defaultValues: {
   colors: []
  }
})
```
Setting `defaultValue` on the input level will only work for uncontrolled inputs _(used with `register` form method)_. In your case, the form field `colors` will be `undefined`, if you set `defaultValue` on the `select` instead of `defaultValues` at the `useForm` hook

**Reply by aruljayaraj** [reply]
@Moshyfawn @bluebill1049 I've tried your solution, it seems working fine with the normal HTML select box, when I work with Ionic version 6 it causes an infinite loop. I've updated the code in the same sandbox URL for your reference. I appreciate it. Thanks.

**Reply by Moshyfawn** [reply]
Alright, it's not an issue with RHF or the Ionic lib. It's that you import RHF hooks from `./src` instead of your actual installed RHF package. See [this working CSB](https://codesandbox.io/s/fervent-bogdan-lzgk9?file=/src/App.tsx)
You want to do
```js
import { useForm, Controller } from "react-hook-form";
```
and you're doing
```js
import { useForm, Controller } from "./src";
```
[The CSB you're using](https://codesandbox.io/s/eager-sun-jt4df) is for debugging and PRs. You want to use [this one](https://codesandbox.io/s/react-hook-form-v7-ts-template-4tn8i) instead.

@bluebill1049 we really need to do smth with [this CSB](https://codesandbox.io/s/eager-sun-jt4df) template. After the recent update, the local RHF version in `./src` causes all sorts of crazy sh*t. I've already closed #7700 with the same issue. And ppl keep using it for, I guess, prototypes and stuff without noticing where the RHF stuff is being imported from and mixing `import {  } from "./src"` and `import {  } from "react-hook-form"`

**Reply by Moshyfawn** [reply]
Actually, @aruljayaraj, I think your issue is also coming from the fact that Ionic sets it's input value internally and.. breaks on a passed `value`. Idk, I'm not familiar with the framework, but if you omit setting `value={field.value}`, the Select component will set it's value on it's own. See [this CSB](https://codesandbox.io/s/fervent-bogdan-lzgk9?file=/src/App.tsx)

**Reply by aruljayaraj** [reply]
@Moshyfawn @bluebill1049 I've updated from `import { } from "./src"` to `import { } from "react-hook-form"`. And as per your suggestion I omitted `value={field.value}` and I can able to get and submit the value. But If I remove that attribute, default values not showing.

**Reply by Moshyfawn** [reply]
Yeah, @aruljayaraj , sry, you're completely right. I didn't think it through, 'cuz I was in a rush.
I seems like some change, that has been introduced with [7.19.0](https://github.com/react-hook-form/react-hook-form/releases?q=7.19.0&expanded=true), is affecting you.
The latest RHF version that works as expected with Ionic input was [7.18.1](https://github.com/react-hook-form/react-hook-form/releases?q=7.18.1&expanded=true). See [this CSB](https://codesandbox.io/s/pedantic-rhodes-wryr9?file=/src/App.tsx). I suggest you lock your RHF version at [7.18.1](https://github.com/react-hook-form/react-hook-form/releases?q=7.18.1&expanded=true) for now.

**Reply by Moshyfawn** [reply]
@aruljayaraj you can track the issue here: #7708

**Reply by aruljayaraj** [reply]
@Moshyfawn I really appreciate it. I've downgraded RHF from **7.25.3** to **7.18.1**. It works as I expected. Thanks a lot, @Moshyfawn @bluebill1049

**Reply by bluebill1049** [reply]
Thanks @Moshyfawn will looking into it.

**Reply by Moshyfawn** [reply]
Awesome! Sry, I couldn't see the issue glancing over the changes between the versions as I didn't have much time today

**Reply by bluebill1049** [reply]
> @bluebill1049 we really need to do smth with this CSB template. After the recent update, the local RHF version in ./src causes all sorts of crazy sh*t. I've already closed #7700 with the same issue. And ppl keep using it for, I guess, prototypes and stuff without noticing where the RHF stuff is being imported from and mixing import { } from "./src" and import { } from "react-hook-form"

updated. thank you

**Reply by bluebill1049** [reply]
Updated with the issue and done with my investigation.

**Reply by aruljayaraj** [reply]
Thanks @Moshyfawn @bluebill1049

**Reply by bluebill1049** [reply]
you are welcome.

**Reply by Moshyfawn** [reply]
Thank you for your time, @bluebill1049
[I've suggested to omit setting the IonSelect `value`](https://github.com/react-hook-form/react-hook-form/discussions/7706#discussioncomment-2089570) to try avoiding the issue, but it caused an issue with setting it's defaultValue instead.
It's interesting, tho, that the issue with Ionic came up with a RHF version changes

@aruljayaraj if you go through with creating an issue with Ionic, you can link this discussion `react-hook-form/react-hook-form#7708` and the issue `react-hook-form/react-hook-form#7706` without the formatting

**Reply by bluebill1049** [reply]
@Moshyfawn we use to return the same reference on each state update within the `useController`, in 7.19.0 we have made `useWatch` reusable within `useControoler`, so each state update with return a new referenced state (shallow clone). My understanding is a component should be resilient for state updates whether it's a shallow clone or not instead of going into the infinite loop.

**Reply by Moshyfawn** [reply]
Got it!
Yeah, Ionic has this problem with loops for different component for some reason. The DateRange component is another one ionic-team/ionic-framework#21101. Not sure what's the motivation there

**Reply by bluebill1049** [reply]
Yea, I saw the stack trace which has a `valueChanged` detection and then `emit` change. I am not sure the motivation behind it, but I can see issues that could happen around such implementation.

**Reply by aruljayaraj** [reply]
I've opened an issue with Ionic Team. Here is the discussion link
https://github.com/ionic-team/ionic-framework/issues/24691

**Reply by npup** [reply]
> For example:
> 
> ```js
> <select
>   multiple
>   onChange={(event) =>
>     onChange(
>       Array.from(event.target.selectedOptions).map(
>         (selectedOption) => selectedOption.value
>       )
>     )
>   }
>   value={value}
> />
> ```

I found this issue/discussion and above solution just as I was about to submit an issue myself (had already made a minimal reproduction at https://github.com/npup/rhf-multiple-select-problem).

The hijacking of onChange works, as stated, but I wonder about the function signatures here.  When using TypeScript, it complains because the `onChange` expects a `ChangeEvent<HTMLSelectElement>` instead of the calculated array of strings.

```
Argument of type 'string[]' is not assignable to parameter of type 'ChangeEvent<HTMLSelectElement>'.
  Type 'string[]' is missing the following properties from type 'ChangeEvent<HTMLSelectElement>': target, nativeEvent, currentTarget, bubbles, and 11 more.
```
What am I missing? What `onChange` is this, that is makes everything work when given the calculated value in the above solution?

Best regards
/p
